packages/vite/src/node/optimizer/esbuildDepPlugin.ts

```diff
@@ -312,19 +315,38 @@ export function esbuildCjsExternalPlugin(externals: string[]): Plugin {
         `^${text.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&')}$`
       const filter = new RegExp(externals.map(escape).join('|'))
 
-      build.onResolve({ filter: /.*/, namespace: 'external' }, (args) => ({
-        path: args.path,
-        external: true
-      }))
+      build.onResolve({ filter: new RegExp(`^${nonFacadePrefix}`) }, (args) => {
+        return {
+          path: args.path.slice(nonFacadePrefix.length),
+          external: true
+        }
+      })
+
+      build.onResolve({ filter }, (args) => {
+        if (args.kind === 'require-call') {
+          return {
+            path: args.path,
+            namespace: cjsExternalFacadeNamespace
+          }
+        }
 
-      build.onResolve({ filter }, (args) => ({
-        path: args.path,
-        namespace: 'external'
-      }))
+        return {
+          path: args.path,
+          external: true
+        }
+      })
 
-      build.onLoad({ filter: /.*/, namespace: 'external' }, (args) => ({
-        contents: `export * from ${JSON.stringify(args.path)}`
```

Comment:
If this is changed to 
```js
`export * from ${JSON.stringify(args.path)}` +
`export { default } from ${JSON.stringify(args.path)}`
```
the following code will be generated.
```js
// external:react
var react_exports = {};
__export(react_exports, {
  default: () => default2
});
import { default as default2 } from "react";
import * as react_star from "react";
var init_react = __esm({
  "external:react"() {
    __reExport(react_exports, react_star);
  }
});
```
This code requires the externalized package (`react` in this example) to have a default export. But we don't know whether the package has a default export. So this code will fail for a package that doesn't have a default export and cannot be used.

Reply:
@sapphi-red @bluwy what about instead doing

```
`export mod from ${JSON.stringify(args.path)}` +
`module.exports = mod`
```

seems esbuild handles mixed ESM and CJS. And this fixes the problem with default exports that I mentioned below and ends up creating a smaller bundle for me as well.

Reply:
That's not correct on two points.

1. `mod` is the value of default export. So named exports will be lost.
1. That import is a self referential. It won't work.

Reply:
@sapphi-red 

> 1. mod is the value of default export. So named exports will be lost.

isn't it the default that is returned when you do `require('module')` which we are targeting with this, is there a way to `require` named exports ? 

> 2. That import is a self referential. It won't work.

not sure about this one, if it is self referential this should not work right ? strangely with this change applied, vite was building properly

Reply:
> isn't it the default that is returned when you do `require('module')` which we are targeting with this, is there a way to `require` named exports ?

In node.js, we can't require a ESM file. In webpack, `require('./esm.js')` returns the module namespace (`Module {__esModule: true, Symbol(Symbol.toStringTag): 'Module', default: 'default', named: 'named'}`).

> not sure about this one, if it is self referential this should not work right ? strangely with this change applied, vite was building properly

I don't know because what you changed is ambiguous. I understood that you changed ``contents: `export * from ${JSON.stringify(args.path)}` `` into that. I rethink that it could work.

Reply:
the patch I had on 3.1.7 is 
```
diff --git a/node_modules/vite/dist/node/chunks/dep-0ebb5606.js b/node_modules/vite/dist/node/chunks/dep-0ebb5606.js
index e51e428..35785cd 100644
--- a/node_modules/vite/dist/node/chunks/dep-0ebb5606.js
+++ b/node_modules/vite/dist/node/chunks/dep-0ebb5606.js
@@ -35222,7 +35222,7 @@ function esbuildCjsExternalPlugin(externals) {
                 namespace: 'external'
             }));
             build.onLoad({ filter: /.*/, namespace: 'external' }, (args) => ({
-                contents: `export * from ${JSON.stringify(args.path)}`
+                contents: `import mod from ${JSON.stringify(args.path)}; module.exports = mod`
             }));
         }
     };

```

but as I said my case was using `ssr` to build lambda functions only externalising node builtins, so I won't be surprised, it this only works for this very special case 🙈